### DISCUSSION TITLE
[help] How to tidy-select target outputs within the command of a downstream target

### DISCUSSION THREAD

**Discussion by shiyw** [ropensci/tarchetypes#178] [Help]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

Intention: in the second `tar_map`, I want to access the outputs of the first `tar_map` with name `a` statically expanded to `a_1` and `a_2`.

```r
list(
  tar_map(
    values = tibble::tibble(x = 1:3),
    tar_target(a, x)
  ),
  tar_map(
    values = tibble::tibble(y = 1:2),
    tar_target(b, a + y)
  )
)
```

However, when `tar_make` triggers the error: `a` does not exist.

I tried a few walkaround.

Those failed:
- `get(paste0("a_", y))` : `a_1` is not in the envir
- `tar_load(paste0("a_", y))` or  `tar_read(paste0("a_", y))` : `tar_load` and `tar_read` are not allowed

Thus, I have to explicitly include these variables so that they can be captured into the enclosure. Like:
```r
list(
  tar_map(
    values = tibble::tibble(x = 1:3),
    tar_target(a, x)
  ),
  tar_map(
    values = tibble::tibble(y = 1:2),
    tar_target(
      b,
      {
        list(a_1, a_2) |> invisible()
        get(paste0("a_", y)) + y
      }
    )
  )
)
```

Wrapping it up:
```r
list(
  tar_map(
    values = tibble::tibble(x = 1:3),
    tar_target(a, x)
  ),
  tar_map(
    values = tibble::tibble(y = 1:2),
    tar_target(
      b,
      {
        tar_use(a_1, a_2)
        tar_get("a", y) + y
        # a <- tar_get2("a")  # factory version
        # a(y) + y  
      }
    )
  )
)
```

with some utils:
```r
# should be tar_sourced
tar_use <- \(...) {}
tar_get <- function(basename, ...) {
  env <- parent.frame()
  get(paste(basename, ..., sep = "_"), envir = env)
}
# factory version
tar_get2 <- function(basename) {
  env <- parent.frame()
  function(...) {
    get(paste(basename, ..., sep = "_"), envir = env)
  }
}
```

My question: Is there any better (tidier) way to collect branching outputs within `command` of `tar_target`? For example, explicitly declare (with `tidy_select`) names by an parameter of `tar_target`.

PS: the required feature may also affect construction of dependency DAG. But currently I only care about the former.

**Comment by wlandau** [comment]
If I understand correctly, you want some branches of target `a`, and you want each branch of target `a` to have two branches of target `b`:

![Screenshot 2024-07-08 at 2 18 25 PM](https://github.com/ropensci/tarchetypes/assets/1580860/8d7868c7-55ce-4a6e-a36a-39288edb4f55)

You can express this with nested `tar_map()`s. The inner `tar_map()` explains how `b` branches from `a`, and then the outer `tar_map()` replicates the whole thing over multiple branches of `a`.

```r
library(targets)
library(tarchetypes)
list(
  tar_map(
    values = tibble::tibble(x = 1:2),
    tar_target(a, x),
    tar_map(
      values = tibble::tibble(y = 1:2),
      tar_target(b, a + y)
    )
  )
)
```

**Reply by shiyw** [reply]
I want some (but not all) branches of target a to have two branches of target b.

**Reply by wlandau** [reply]
In that case, you would need the full flexibility of something like `tar_eval()` (c.f. https://books.ropensci.org/targets/static.html#metaprogramming) or manually write out each target without static branching.

```r
library(targets)
library(tarchetypes)
list(
  tar_eval(
    values = list(a = rlang::syms(c("a_1", "a_2"))),
    tar_target(a, x)
  ),
  tar_eval(
    values =  list(
      a = rlang::syms(c("a_1", "a_1", "a_2")),
      b = rlang::syms(c("b_1", "b_2", "b_3"))
    ),
    tar_target(b, a + y)
  )
)
```

![Screenshot 2024-07-16 at 3 05 54 PM](https://github.com/user-attachments/assets/f488cb17-c2b5-4dc6-b463-6c02b2c6aae1)
